ztunnel/xds: Filter workloads by enrolled namespace

nddq opened the pull request:
## Summary

Fix ztunnel traffic handling between enrolled and unenrolled workloads by filtering xDS workload data based on namespace enrollment state.

- **Namespace filtering**: Only workloads in enrolled namespaces are sent to ztunnel via xDS. Previously all workloads were sent regardless of enrollment status, causing issues with mixed-enrollment traffic.
- **PodUID in CES**: Add `PodUID` field to `CoreCiliumEndpoint` so CES-derived endpoints carry the Pod UID. Without this, all CES workloads are sent to ztunnel with an empty UID, so ztunnel treats them as the same workload and only retains the last one received.
- **Remove `ListAllEndpoints`**: Replace the initial snapshot + subscribe pattern with resource event replay via `SubscribeToEndpointEvents`. This eliminates double-sending all workloads on connect and avoids potential deadlocks from blocking sends on a bounded channel while the event loop is waiting.
- **Configurable buffer**: Make `endpointEventChan` buffer size configurable via `--ztunnel-endpoint-event-channel-buffer-size` (default 1). Mark the flag as hidden.
- **Reconnect safety**: Add context guards in `emitEndpointEvents` and `SubscribeToEndpointEvents` to avoid sending on a stale channel after stream cancellation.
- **E2E tests**: Add enrolled-to-unenrolled and unenrolled-to-enrolled connectivity test scenarios (same-node and different-node) to verify plain text traffic between mixed enrollment namespaces.

Fixes: #43113

```release-note
ztunnel: Fix traffic handling between enrolled and unenrolled workloads by filtering xDS data based on namespace enrollment state.
```

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -190,11 +217,29 @@ func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, wg *syn
+					logfields.Name, e.Object.GetName(),
+				)
+				break
+			}
```
I think this can be done outside the switch statement to avoid duplication? Same for handleCESUpsert/handleCESDelete I believe.

Also consider making a small util function for the lookup, I think that would help with readability of the code.

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -208,12 +253,22 @@ func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, wg *syn
 
 func (es *EndpointSource) ListAllEndpoints(ctx context.Context) ([]*types.CiliumEndpoint, error) {
```
Isn't this function obsolete now? Can't we rely on the fact that the reconciler will send all necessary endpoints based on the enrolled namespace table? Not sure if that's a bad thing but keeping ListAllEndpoints means we will definitely send all workloads to xds server of ztunnel twice I think.

nddq replied:
I believe this still applies at runtime when `handleDeltaDiscoveryReq` is called after ztunnel reconnects to the cilium agent. In that scenario, ztunnel needs to repopulate its state. If all namespaces are already enrolled, the reconciler won't send any updates, so this code path becomes necessary for state recovery.

nddq replied:
just retraced the code again and you are right, `SubscribeToEndpointEvents` will replay all existing resources as Upsert events on ztunnel's reconnection, i'll remove this

rgo3 commented on pkg/ztunnel/xds/xds_server.go:
```diff
@@ -79,13 +85,21 @@ type Server struct {
 		log:                       log,
 		k8sCiliumEndpointsWatcher: k8sCiliumEndpointsWatcher,
 		epManager:                 EPManager,
+		endpointEventChan:         make(chan *EndpointEvent, 1024),
```
Is there any concern with this buffer size? There are a few blocking sends on this channel, so if there are namespaces with more than 1024 pods we might see scalability issues.

nddq replied:
Hard to say since I don't have the scale test numbers yet, but we can always make this configurable.

nddq replied:
i've added a config field to make this configurable, defaulting to 1024

rgo3 reviewed (changes requested):
Gave this a first pass, mostly looks good.

Some general concerns with sending double updates (although I think ztunnel can handle this) and the buffer capacity of the update channel.

We might want to let @ldelossa or @hemanthmalla take another look next week though if this needs to make the feature freeze as I'm out for the rest of the year.

nebril reviewed (approved):
k8s LGTM

joestringer commented on pkg/ztunnel/config/config.go:
```diff
@@ -8,16 +8,19 @@ import (
 
 func (c Config) Flags(flags *pflag.FlagSet) {
 	flags.Bool("enable-ztunnel", false, "Use zTunnel as Cilium's encryption infrastructure")
+	flags.Int("ztunnel-endpoint-event-channel-buffer-size", 1024, "Buffer size for the ztunnel endpoint event channel")
```
I strongly suggest hiding this flag (`MarkHidden`). We have far too many flags in the agent already, and realistically users are not going to understand how to set a reasonable value for this. If we're currently figuring out a reasonable default as part of the ztunnel beta then we can keep the flag hidden and test out different values to understand the behavior. If we really think there is some sort of tuning the user should do then we could consider whether to expose it later on, but my guess is we just need to find some reasonable default or _maybe_ if we're getting fancy, have some heuristic to pick.

joestringer replied:
side note also, I'd suggest testing with buffer size 0 or 1 to confirm that there's no weird behavior in those cases (such as deadlocks from blocking when the buffer is full)

nddq replied:
gotcha, will make the change and test that out

nddq replied:
run a local test with some pod churns, no weird behaviors observed

giorio94 replied:
I also second Joe's concern about using a channel with a a buffer greater than one for caching, which is often a source of problems down the line, that only become evident at a scale large enough to cause the buffer to fill. What is the reason for not hard-coding a buffer size of either 0 or 1?

nddq replied:
Might be a fair suggestion since the channel only has one consumer, @rgo3 what do you think?

rgo3 replied:
I think IIRC the original reason to make it 1024 was to ensure batching in `func (sp *StreamProcessor) handleEPEvent(epEvent *EndpointEvent) error`. This function drains the endpoint updates and can batch them in a single XDS update. I think with a buffer size of 0 the batching would be broken, with 1 it could still work because it gives the reconciler more time to write a new event to the channel? But I couldn't say for certain.

rgo3 replied:
Hm, I don't really know what's best here. Let's make the default 1, and see where we end up with that. We should also add some validation that the value can't be negative.

giorio94 reviewed (approved):
Approving for the two small changes in the `ciliumendpointslice` package.

Overall, I think that this commit is trying to address too may problems at once, and it is a good candidate for being split into multiple separate ones (for instance, one or more for each of the bullet points in your description), each focusing on a specific task. That would simplify the review process, and also bisecting possible issues in the future. However, I won't block on this if the other reviewers are fine with this approach.

MikeZappa87 commented on pkg/ztunnel/xds/xds_server.go:
```diff
@@ -71,6 +74,11 @@ type Server struct {
+	enrolledNamespaceTable    statedb.RWTable[*table.EnrolledNamespace]
+	// endpointEventChan is shared across streams. Overlapping sends during
+	// ztunnel reconnect are safe since xDS upserts are idempotent.
+	endpointEventChan chan *EndpointEvent
```
Not sure if cilium has a style guide, should this be tabbed over (or spaced for you spaces folks)

joestringer replied:
If `go fmt` is happy, we're happy.

MikeZappa87 replied:
If it fits, it ships! Right on, I am assuming go fmt has already been run then

nddq replied:
yep, vscode autoruns that on file save

nddq committed:
ztunnel/xds: Add PodUID to CoreCiliumEndpoint

Add PodUID field to CoreCiliumEndpoint so CES-derived endpoints carry
the Pod UID. Without this, all CES workloads are sent to ztunnel with
an empty UID, so ztunnel treats them as the same workload and only
retains the last one received.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

nddq committed:
ztunnel/xds: Make endpoint event channel configurable and shared

Make endpointEventChan buffer size configurable via
--ztunnel-endpoint-event-channel-buffer-size (default 1). Mark the
flag as hidden. Add validation rejecting negative values.

Share a single endpointEventChan across streams instead of creating a
per-stream channel. Overlapping sends during ztunnel reconnect are safe
since xDS upserts are idempotent.

Change NewServer from cell.Invoke to cell.Provide so that the shared
channel can be injected into other cells.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

nddq committed:
ztunnel/xds: Buffer replayed events for atomic initial snapshot

Remove ListAllEndpoints and the sync.WaitGroup-based subscription in
favour of a two-phase SubscribeToEndpointEvents that delivers all
existing workloads as a single atomic xDS response.

Phase 1 (pre-Sync): buffer all replayed endpoint events from the
Kubernetes resource framework into a local EndpointEventCollection.
On Sync: send the buffered collection to handleAddressTypeURL via a
dedicated syncCh channel.
Phase 2 (post-Sync): forward new events to endpointRecv individually
as before.

handleAddressTypeURL now blocks on syncCh to receive the initial batch,
then sends it as one DeltaDiscoveryResponse. This eliminates the
double-sending of workloads on connect and prevents the race where
ztunnel marks the Address type as initialized with zero workloads then
receives individual endpoint updates. Since ztunnel blocks processing
ZDS updates until it has received the initial Address snapshot,
delivering the full set atomically avoids a window where pod traffic
could be misrouted.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

nddq committed:
ztunnel/xds: Add context guards for reconnect safety

Add context guards in emitEndpointEvents and SubscribeToEndpointEvents
to avoid sending on a stale channel after stream cancellation. This
prevents goroutines from blocking indefinitely when the stream context
is canceled during a ztunnel reconnect.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

nddq committed:
ztunnel/xds: Filter workloads by enrolled namespace

Only workloads in enrolled namespaces are sent to ztunnel via xDS.
Previously all workloads were sent regardless of enrollment status,
causing issues with mixed-enrollment traffic.

The enrollment reconciler now sends existing CiliumEndpoints and
CiliumEndpointSlices to the xDS server when a namespace is enrolled,
and sends delete events when a namespace is unenrolled. A NamespaceIndex
is added to both CiliumEndpoint and CiliumEndpointSlice resource stores
to enable efficient namespace-scoped lookups.

Introduce CiliumEndpointsWatcher interface to decouple the stream
processor from the concrete K8sCiliumEndpointsWatcher type, enabling
mock-based testing.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

giorio94 commented on pkg/k8s/apis/cilium.io/v2alpha1/types.go:
```diff
@@ -36,6 +36,9 @@ type CoreCiliumEndpoint struct {
 	// IdentityID is the numeric identity of the endpoint
 	// +kubebuilder:validation:Optional
 	IdentityID int64 `json:"id,omitempty"`
+	// PodUID is the UID of the Pod that owns this endpoint.
+	// +kubebuilder:validation:Optional
+	PodUID string `json:"pod-uid,omitempty"`
```
The `CustomResourceDefinitionSchemaVersion` constant should be bumped given you are updating the CRD schema.

giorio94 commented on pkg/k8s/factory_functions.go:
```diff
@@ -213,20 +215,42 @@ func ConvertCEPToCoreCEP(cep *cilium_v2.CiliumEndpoint) *cilium_v2alpha1.CoreCil
+
 // ConvertCoreCiliumEndpointToTypesCiliumEndpoint converts CoreCiliumEndpoint object to types.CiliumEndpoint.
 func ConvertCoreCiliumEndpointToTypesCiliumEndpoint(ccep *cilium_v2alpha1.CoreCiliumEndpoint, ns string) *types.CiliumEndpoint {
+	var ownerRefs []slim_metav1.OwnerReference
+	if ccep.PodUID != "" {
+		ownerRefs = []slim_metav1.OwnerReference{
+			{
+				Kind: "Pod",
+				UID:  k8sTypes.UID(ccep.PodUID),
+			},
+		}
+	}
```
Does this need to be the Pod UID, or could it be generalized to simply be an UID, and then directly propagate the one of the CiliumEndpoint, to avoid having to rely on the owner reference? 

Which approach are you using in ztunnel in case of vanilla CiliumEndpoints (I assume you are not using the Pod UID in that case, as not part of the CiliumEndpoint CRD)?

nddq replied:
We need the pod UID down in ztunnel for PID-based workload attestation down the line. For CiliumEndpoint, we are getting the pod UID via the OwnerReference

nddq replied:
https://github.com/cilium/cilium/pull/43388

giorio94 replied:
ACK, thanks for confirming.

giorio94 commented on pkg/k8s/factory_functions_test.go:
```diff
@@ -1104,7 +1105,7 @@ func Test_TransformToCiliumEndpoint(t *testing.T) {
 							Generation:      5,
 							CreationTimestamp: metav1.Time{
-								Time: time.Date(2018, 01, 01, 01, 01, 01, 01, time.UTC),
+								Time: time.Date(2018, 0o1, 0o1, 0o1, 0o1, 0o1, 0o1, time.UTC),
```
Please revert this unnecessary change.

nddq replied:
Might be a formatting on save thing, will revert this

giorio94 reviewed (commented):
Thanks for splitting the PR in multiple commits, it is now way easier to focus on the individual changes. I've only looked at the first commit pertaining to the CES changes, and left a couple of comments/questions inline.

giorio94 reviewed (approved):
Thanks. The first commit looks good to me.

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -173,11 +181,36 @@ func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, wg *syn
 				continue
 			}
 
-			switch e.Kind {
```
The new lines you are adding here are reusing the logic implemented in `handleCESUpsert/Delete`, so I think you could keep this code quite a bit simpler by refactoring `handleCESUpsert/Delete` into returning and `EndpointEventCollection`.

E.g.:
```Go
var events EndpointEventCollection
switch e.Kind {
case resource.Upsert:
	events = es.handleCESUpsert(e.Object, cesCache, e.Key)
case resource.Delete:
	events = es.handleCESDelete(e.Object, cesCache, e.Key)
}

if !synced {
	initialBatch = append(initialBatch, events...)
} else {
	events.EmitTo(es.sp.stream, es.sp.endpointRecv)
}
```

In this approach `func (c EndpointEventCollection) EmitTo` would replace `func (es *EndpointSource) emitEndpointEvents`.

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -187,60 +220,62 @@ func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, wg *syn
+			switch e.Kind {
+			case resource.Upsert:
+				es.sp.endpointRecv <- &EndpointEvent{
+					Type:           CREATE,
+					CiliumEndpoint: e.Object,
+				}
+			case resource.Delete:
+				es.sp.endpointRecv <- &EndpointEvent{
+					Type:           REMOVED,
+					CiliumEndpoint: e.Object,
+				}
 			}
 		}
```
A slightly less verbose approach:
```Go
var eventType EndpointEventType
switch e.Kind {
case resource.Upsert:
	eventType = CREATE
case resource.Delete:
	eventType = REMOVED
}

event := &EndpointEvent{Type: eventType, CiliumEndpoint: e.Object}
if !synced {
	initialBatch = append(initialBatch, event)
} else {
	es.sp.endpointRecv <- event
}

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -343,15 +296,12 @@ func TestStreamProcessorDeltaDiscoveryRequest(t *testing.T) {
-		if _, ok := sp.expectedNonce[req.ResponseNonce]; ok {
-			t.Fatalf("Expected nonce %q to be removed, but it still exists", req.ResponseNonce)
-		}
+		_, ok := sp.expectedNonce["x"]
```
```
_, ok := sp.expectedNonce[req.ResponseNonce]
```
In case the request is ever changed it's better to use the value from the previous declared variable.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -370,147 +320,229 @@ func TestStreamProcessorDeltaDiscoveryRequest(t *testing.T) {
+		require.NoError(t, err, "Nack should not return error, just log it")
+
+		// Verify nonce was removed
+		_, ok := sp.expectedNonce["nack-nonce"]
```
Same as above.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -318,13 +250,34 @@ func TestStreamProcessorEndpointEvents(t *testing.T) {
 		}
 	})
+
+	t.Run("SendMsg error returns error", func(t *testing.T) {
```
This test just tests if `handleEPEvent` can forward an error from a function it's calling, right? I don't think tests like these are necessary, I trust that a return statement in Go does its job ;)

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -370,147 +320,229 @@ func TestStreamProcessorDeltaDiscoveryRequest(t *testing.T) {
+		require.True(t, ok)
+	})
+
+	t.Run("returns error when SendMsg fails", func(t *testing.T) {
```
Again, I don't think this tests any actual business logic and I trust that an error get bubbled up if we return it.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -370,147 +320,229 @@ func TestStreamProcessorDeltaDiscoveryRequest(t *testing.T) {
 		require.Len(t, added, 1)
 		require.Len(t, updated, 1)
 		require.Len(t, removed, 1)
-		require.Equal(t, "pod-3", added[0].Name)
```
I think we should keep these Equal checks. While unlikely, if we introduce a bug that adds to the wrong slice but still adds one element to each one we don't test the full expectation anymore.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -370,147 +320,229 @@ func TestStreamProcessorDeltaDiscoveryRequest(t *testing.T) {
+		require.Equal(t, "0", capturedResponse.Nonce)
+
+		// Verify nonce was tracked
+		_, ok := sp.expectedNonce["0"]
```
See above.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -85,6 +101,30 @@ func (m *MockEndpointEventSource) GetSubscriptionStatus() bool {
+	}
+	return batch
+}
+
```
These two functions are identical, so they can be merged into one, e.g. `subscribeToEndpoints`

rgo3 commented on pkg/ztunnel/reconciler/reconciler.go:
```diff
@@ -51,18 +63,58 @@ func NewEnrollmentReconciler(cfg params) reconciler.Operations[*table.EnrolledNa
+				}
+			}
+		}
+	} else {
```
Nit, an early return in the above `if` eliminates the need for `else`

rgo3 commented on pkg/ztunnel/reconciler/reconciler.go:
```diff
@@ -84,6 +136,42 @@ func (ops *EnrollmentReconciler) Update(ctx context.Context, txn statedb.ReadTxn
+				CiliumEndpoint: cep,
+			}
+		}
+	}
```
The body of this function is nearly identical to the Update function, please extract a shared helper that takes the EndpointEventType as an argument.

rgo3 commented on pkg/ztunnel/reconciler/reconciler_test.go:
```diff
@@ -376,11 +531,119 @@ func TestEnrollmentReconciler_Update(t *testing.T) {
 			expectedError:    false,
 		},
+		{
+			name:                   "no endpoints in namespace",
```
Duplicate of the test-case above?

rgo3 commented on pkg/ztunnel/reconciler/reconciler_test.go:
```diff
@@ -376,11 +531,119 @@ func TestEnrollmentReconciler_Update(t *testing.T) {
+			expectedError:          false,
+		},
+		{
+			name:                      "enroll with multiple CiliumEndpointSlices",
```
Does this case test any special case scenario in our code except that looping in Go works compared to the case above? If it's only testing the for loop we can remove it IMO.

rgo3 commented on pkg/ztunnel/reconciler/reconciler_test.go:
```diff
@@ -398,20 +661,128 @@ func TestEnrollmentReconciler_Update(t *testing.T) {
+		assert.Contains(t, err.Error(), "failed to get CiliumEndpointSlice store")
+	})
+}
+
```
I think these tests again just test that error forwarding works in Go, so I'd be in favor of removing them.

nddq replied:
sounds good, i was trying to max out the code coverage 😅 Will revise the tests to see if I can remove anything else as well if they don't test the functionality directly

rgo3 commented on pkg/ztunnel/reconciler/reconciler_test.go:
```diff
@@ -477,6 +910,26 @@ func TestEnrollmentReconciler_Delete(t *testing.T) {
+					time.Sleep(10 * time.Millisecond)
+				}
+			}
+			assert.Equal(t, tt.expectedEndpointEvents, eventCount, "unexpected number of endpoint events")
```
This code is identical to the other drainLoop label code. Please extract a common helper that takes the EndpointEventType.

rgo3 commented on pkg/ztunnel/xds/stream_processor_test.go:
```diff
@@ -708,8 +760,43 @@ func TestHandleCESUpsert(t *testing.T) {
+	})
+
+}
+
```
These tests aren't really testing functionality of the StreamProcessor, and this pkg doesn't need to add tests to cover stateDB code. I'd assume calling Get on a table is covered 😅

rgo3 commented on pkg/ztunnel/reconciler/reconciler.go:
```diff
@@ -51,18 +63,58 @@ func NewEnrollmentReconciler(cfg params) reconciler.Operations[*table.EnrolledNa
+		for _, ces := range slices {
+			for _, coreCep := range ces.Endpoints {
+				cep := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&coreCep, ces.Namespace)
+				ops.endpointEventCh <- &xds.EndpointEvent{
```
Do we need to wrap those blocking channel writes into a select that watches the context passed to this function?

rgo3 reviewed (changes requested):
For details check my comments inline, but I think we can shrink the LOC of this PR a bit by extracting some helpers of duplicated code and removing tests that don't really test functionality added in this PR.

My main technical concern are some of the blocking channel sends if no ztunnel is connected. I think we can be defensive here and add some select statements to also watch the context that is passed to the functions doing those sends. A counter argument could be that without a ztunnel connection it doesn't really matter if the reconciler is blocked. WDYT? 

Thanks for your patience @nddq, I can see light at the end of the tunnel 😅

rgo3 commented on pkg/k8s/resource_ctors.go:
```diff
@@ -391,6 +391,7 @@ func CiliumSlimEndpointResource(params CiliumResourceParams, localNodeStore *nod
 		opts...,
 	)
 	indexers := cache.Indexers{
+		NamespaceIndex: namespaceIndexFunc,
```
I think this doesn't work. `namespaceIndexFunc` calls object.GetNamespace() which returns meta.Namespace. But a CES resource doesn't have meta.Namespace set because it's a cluster scoped resource I believe the actual namespace lives in the spec-level ces.Namespace field. Did you test with endpoint slices?

nddq replied:
Good catch! I spun up a CES-enabled cluster and verified this. I've added a dedicated index func that reads the spec-level Namespace field instead. Not sure why I didn't catch this since I do remember testing with CES at some point, but probably to do with the order of operation: if the namespace is enrolled before pods come up, workloads arrive through `SubscribeToEndpointEvents` (which reads `e.Object.Namespace` directly and works fine). The bug only manifests when the namespace is enrolled after pods already exist, since that hits the reconciler's `emitEndpointEvents` path which queries the broken index.

nddq replied:
The fix is down below, looks like you commented on the wrong resource 🙂

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -252,31 +287,33 @@ func (sp *StreamProcessor) handleAddressTypeURL(req *v3.DeltaDiscoveryRequest) e
+	// events until the initial Sync, then sends the batch on syncCh.
+	// Post-Sync events are forwarded to endpointRecv individually.
+	syncCh := make(chan EndpointEventCollection, 1)
+	go sp.endpointSource.SubscribeToEndpointEvents(ctx, syncCh)
```
Do we need to guard against handleAddressTypeURL potentially being called multiple times? I know this behavior is not introduced in this PR specifically but it reads to me as if we could potentially spawn a new goroutine evertime ztunnel sends this type of request. Can this happen? cc @ldelossa

nddq replied:
I feel like that would be a ztunnel's bug if that happen

nddq replied:
Doesn't hurt to add a simple guard though

rgo3 replied:
Let's leave this as a follow up, I'll create an issue to track this.

rgo3 replied:
#44746

rgo3 commented on pkg/ztunnel/xds/stream_processor.go:
```diff
@@ -189,36 +203,27 @@ func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, syncCh
 				continue
 			}
 
+			if !es.isNamespaceEnrolled(e.Object.Namespace) {
```
Nit, but this change was probably supposed to go into commit 3 😬

nddq replied:
Let me do a re-rebase 🥲

nddq replied:
actually, this still might make more sense for commit 4 since commit 3 only add the context guards IMO, let me know if this is a dealbreaker

rgo3 commented on pkg/ztunnel/reconciler/reconciler.go:
```diff
@@ -84,6 +153,10 @@ func (ops *EnrollmentReconciler) Update(ctx context.Context, txn statedb.ReadTxn
 }
 
 func (ops *EnrollmentReconciler) Delete(ctx context.Context, txn statedb.ReadTxn, rev statedb.Revision, ns *table.EnrolledNamespace) error {
+	if err := ops.emitEndpointEvents(ctx, ns.Name, xds.REMOVED); err != nil {
```
I think on Delete it could make sense to emit the xDS update after we've removed the endpoint from zds? If zds delete succeeds but xds delete fails, the endpoint is still technically not enrolled to ztunnel anymore because the iptables rules are gone. So traffic can continue to flow.
But with the current order, if the xds delete succeeds but the zds delete fails, traffic will start to drop.

nddq replied:
that's makes sense, basically reversing the order from `Update` (adding a workload to the mesh first before redirecting its traffic)

rgo3 commented on pkg/ztunnel/reconciler/reconciler_test.go:
```diff
@@ -301,31 +445,62 @@ func setupTest(t *testing.T) (*statedb.DB, statedb.RWTable[*table.EnrolledNamesp
+
+// drainEndpointEvents drains the endpoint event channel and verifies the expected
+// count and event type.
+func drainEndpointEvents(t *testing.T, ch <-chan *xds.EndpointEvent, expectedCount int, expectedType xds.EndpointEventType) {
```
Let's use `require.Eventually` here instead of the 100ms timeout to avoid any potential flakes.

```Go
func drainEndpointEvents(t *testing.T, ch <-chan *xds.EndpointEvent, expectedCount int, expectedType xds.EndpointEventType) {
        t.Helper()
        var events []*xds.EndpointEvent
        require.Eventually(t, func() bool {
                for {
                        select {
                        case event := <-ch:
                                events = append(events, event)
                        default:
                                return len(events) >= expectedCount
                        }
                }
        }, time.Second, 10*time.Millisecond, "expected %d endpoint events, got %d", expectedCount, len(events))

        require.Len(t, events, expectedCount, "unexpected number of endpoint events")
        for _, event := range events {
                assert.Equal(t, expectedType, event.Type)
        }
}
```

rgo3 reviewed (changes requested):
Last two improvements, otherwise LGTM. Will approve when addressed

nddq committed:
ztunnel/xds: Add enrolled/unenrolled E2E connectivity tests

Add enrolled-to-unenrolled and unenrolled-to-enrolled connectivity test
scenarios (same-node and different-node) to verify plain text traffic
between mixed enrollment namespaces.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

rgo3 reviewed (approved):
Thanks for the patience @nddq
